include/wlr/types/wlr_foreign_toplevel_management_v1.h

```diff
@@ -0,0 +1,112 @@
+	struct wl_global *global;
+	struct wl_list resources;
+	struct wl_list toplevels; // wlr_foreign_toplevel_handle_v1::link
+
+	struct wl_listener display_destroy;
+
+	struct {
+		struct wl_signal destroyed;
+	} events;
+
+	void *data;
+};
+
+enum wlr_foreign_toplevel_handle_state {
+	WLR_FOREIGN_TOPLEVEL_STATE_MAXIMIZED = 1,
+	WLR_FOREIGN_TOPLEVEL_STATE_MINIMIZED = 2,
+	WLR_FOREIGN_TOPLEVEL_STATE_ACTIVATED = 4,
+};
+
+struct wlr_foreign_toplevel_handle_output {
+	struct wl_list link; // wlr_foreign_toplevel_handle_v1::outputs
+	struct wlr_output *output;
+};
+
+struct wlr_foreign_toplevel_handle_v1 {
+	struct wl_list resources;
+	struct wl_list link;
+
+	char *title;
+	char *app_id;
+	struct wl_list outputs; // wlr_foreign_toplevel_output
+	uint32_t state; // wlr_foreign_toplevel_state
+
+	struct {
+		struct wl_signal request_maximize;
+		struct wl_signal request_minimize;
+		struct wl_signal request_activate;
+		struct wl_signal request_set_rectangle;
+		struct wl_signal request_close;
+	} events;
+
+	void *data;
+};
+
+struct wlr_foreign_toplevel_handle_maximized_event {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel;
+	bool maximized;
+};
+
+struct wlr_foreign_toplevel_handle_minimized_event {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel;
+	bool minimized;
+};
+
+struct wlr_foreign_toplevel_handle_activated_event {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel;
+	struct wlr_seat *seat;
+};
+
+struct wlr_foreign_toplevel_handle_set_rectangle_event {
+	struct wlr_foreign_toplevel_handle_v1 *toplevel;
+	struct wl_surface *surface;
+	int32_t x, y, width, height;
+};
+
+struct wlr_foreign_toplevel_manager_v1* wlr_foreign_toplevel_manager_v1_create(
+	struct wl_display *display);
+void wlr_foreign_toplevel_manager_v1_destroy(
+	struct wlr_foreign_toplevel_manager_v1 *manager);
+
+struct wlr_foreign_toplevel_handle_v1*
+wlr_foreign_toplevel_manager_v1_create_toplevel_handle(
+	struct wlr_foreign_toplevel_manager_v1 *manager);
+void wlr_foreign_toplevel_handle_v1_destroy(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel);
+
+void wlr_foreign_toplevel_handle_v1_set_title(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, const char *title);
+void wlr_foreign_toplevel_handle_v1_set_app_id(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, const char *app_id);
+
+void wlr_foreign_toplevel_handle_v1_output_enter(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, struct wlr_output *output);
+void wlr_foreign_toplevel_handle_v1_output_leave(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, struct wlr_output *output);
+
+void wlr_foreign_toplevel_handle_v1_set_maximized(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, bool maximized);
+void wlr_foreign_toplevel_handle_v1_set_minimized(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, bool minimized);
+void wlr_foreign_toplevel_handle_v1_set_activated(
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, bool activated);
+
+void wlr_foreign_toplevel_handle_v1_send_done(
```

Comment:
This should not be exposed. Instead, add a idle event source when changing a value and send `done` when it's fired.

Reply:
oh, that's even better. Hadn't noticed we do things like that.